File: tests/org/elixir_lang/reference/callable/Issue480Test.java
package org.elixir_lang.reference.callable;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiPolyVariantReference;
import com.intellij.psi.PsiReference;
import com.intellij.psi.ResolveResult;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import kotlin.ranges.IntRange;
import org.elixir_lang.NameArityRange;
import org.elixir_lang.psi.CallDefinitionClause;
import org.elixir_lang.psi.call.Call;
import org.jetbrains.annotations.NotNull;

import static org.elixir_lang.psi.CallDefinitionClause.nameArityRange;

public class Issue480Test extends BasePlatformTestCase {
    /*
     * Tests
     */

    public void testLocal() {
        myFixture.configureByFiles("local.ex", "referenced.ex");
        assertReferenceAndResolvedNameArityRange("changeset", 2);
    }

    public void testRemote() {
        myFixture.configureByFiles("remote.ex", "referenced.ex");
        assertUnresolvableReferenceNameArityRange("changeset", 1);
    }

    public void testDirectModuleQualifier() {
        myFixture.configureByFiles("direct_module_qualifier.ex", "referenced.ex");
        assertReferenceAndResolvedNameArityRange("changeset", 1);
    }

    public void testAliasedModuleQualifier() {
        myFixture.configureByFiles("aliased_module_qualifier.ex", "referenced.ex");
        assertReferenceAndResolvedNameArityRange("changeset", 1);
    }

    public void testDoubleAliasesModuleQualifier() {
        myFixture.configureByFiles("double_aliased_module_qualifier.ex", "referenced.ex");
        assertUnresolvableReferenceNameArityRange("changeset", 1);
    }

    public void testMapAccessQualifier() {
        myFixture.configureByFiles("map_access_qualifier.ex", "referenced.ex");
        assertUnresolvableReferenceNameArityRange("__struct__", 1);
    }

    public void testUnresolvedAliasQualifier() {
        myFixture.configureByFiles("unresolved_alias_qualifier.ex", "referenced.ex");
        assertUnresolvableReferenceNameArityRange("changeset", 1);
    }

    /*
     * Protected Instance Methods
     */

    @Override
    protected String getTestDataPath() {
        return "testData/org/elixir_lang/reference/callable/issue_480";
    }

    /*
     * Private Instance Methods
     */

    private void assertUnresolvableReferenceNameArityRange(@NotNull String name, int arity) {
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        Call grandParentCall = (Call) grandParent;

        assertEquals(name, grandParentCall.functionName());
        assertEquals(arity, grandParentCall.resolvedFinalArity());

        PsiReference reference = grandParent.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNull(resolved);
    }

    private void assertReferenceAndResolvedNameArityRange(@NotNull String name, int arity) {
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        Call grandParentCall = (Call) grandParent;

        assertEquals(name, grandParentCall.functionName());
        assertEquals(arity, grandParentCall.resolvedFinalArity());

        assertResolvedNameArityRange(grandParentCall, name, arity);
    }

    private void assertResolvedNameArityRange(@NotNull PsiElement element, @NotNull String name, int arity) {
        PsiReference reference = element.getReference();

        assertNotNull(reference);

        assertInstanceOf(reference, PsiPolyVariantReference.class);

        PsiPolyVariantReference psiPolyVariantReference = (PsiPolyVariantReference) reference;

        ResolveResult[] resolveResults = psiPolyVariantReference.multiResolve(false);

        for (ResolveResult resolveResult : resolveResults) {
            if (resolveResult.isValidResult()) {
                PsiElement resolved = resolveResult.getElement();

                assertInstanceOf(resolved, Call.class);

                Call maybeDefCall = (Call) resolved;

                assertTrue(CallDefinitionClause.is(maybeDefCall));

                assertEquals(new NameArityRange(name, new IntRange(arity, arity)), nameArityRange(maybeDefCall));
            }
        }
    }
}

File: tests/org/elixir_lang/reference/callable/Issue480Test.kt
package org.elixir_lang.reference.callable

import org.elixir_lang.psi.CallDefinitionClause.`is`
import org.elixir_lang.psi.CallDefinitionClause.nameArityInterval
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiPolyVariantReference
import org.elixir_lang.NameArityInterval
import org.elixir_lang.psi.ArityInterval
import com.intellij.psi.ResolveState
import org.elixir_lang.PlatformTestCase
import org.elixir_lang.psi.call.Call

class Issue480Test : PlatformTestCase() {
    fun testLocal() {
        myFixture.configureByFiles("local.ex", "referenced.ex")
        assertReferenceAndResolvedNameArityRange("changeset", 2)
    }

    fun testRemote() {
        myFixture.configureByFiles("remote.ex", "referenced.ex")
        assertUnresolvableReferenceNameArityRange("changeset", 1)
    }

    fun testDirectModuleQualifier() {
        myFixture.configureByFiles("direct_module_qualifier.ex", "referenced.ex")
        assertReferenceAndResolvedNameArityRange("changeset", 1)
    }

    fun testAliasedModuleQualifier() {
        myFixture.configureByFiles("aliased_module_qualifier.ex", "referenced.ex")
        assertReferenceAndResolvedNameArityRange("changeset", 1)
    }

    fun testDoubleAliasesModuleQualifier() {
        myFixture.configureByFiles("double_aliased_module_qualifier.ex", "referenced.ex")
        assertReferenceAndResolvedNameArityRange("changeset", 1)
    }

    fun testMapAccessQualifier() {
        myFixture.configureByFiles("map_access_qualifier.ex", "referenced.ex")
        assertUnresolvableReferenceNameArityRange("__struct__", 1)
    }

    fun testUnresolvedAliasQualifier() {
        myFixture.configureByFiles("unresolved_alias_qualifier.ex", "referenced.ex")
        assertReferenceAndResolvedNameArityRange("changeset", 1)
    }

    override fun getTestDataPath(): String = "testData/org/elixir_lang/reference/callable/issue_480"

    private fun assertUnresolvableReferenceNameArityRange(name: String, arity: Int) {
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)
        val grandParentCall = grandParent as Call

        assertEquals(name, grandParentCall.functionName())
        assertEquals(arity, grandParentCall.resolvedFinalArity())

        val reference = grandParent.getReference()
        assertNotNull(reference)

        val resolved = reference!!.resolve()
        assertNull(resolved)
    }

    private fun assertReferenceAndResolvedNameArityRange(name: String, arity: Int) {
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)
        val grandParentCall = grandParent as Call

        assertEquals(name, grandParentCall.functionName())
        assertEquals(arity, grandParentCall.resolvedFinalArity())
        assertResolvedNameArityRange(grandParentCall, name, arity)
    }

    private fun assertResolvedNameArityRange(element: PsiElement, name: String, arity: Int) {
        val reference = element.reference
        assertNotNull(reference)
        assertInstanceOf(reference, PsiPolyVariantReference::class.java)
        val psiPolyVariantReference = reference as PsiPolyVariantReference

        val resolveResults = psiPolyVariantReference.multiResolve(false)
        for (resolveResult in resolveResults) {
            if (resolveResult.isValidResult) {
                val resolved = resolveResult.element
                assertInstanceOf(resolved, Call::class.java)
                val maybeDefCall = resolved as Call?
                assertTrue(`is`(maybeDefCall!!))
                assertEquals(
                        NameArityInterval(name, ArityInterval(arity, arity)),
                        nameArityInterval(maybeDefCall, ResolveState.initial())
                )
            }
        }
    }
}
